applications/FluidDynamicsApplication/custom_utilities/fluid_adjoint_utilities.h

```diff
@@ -0,0 +1,191 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Suneth Warnakulasuriya
+//
+
+#if !defined(KRATOS_FLUID_ADJOINT_UTILITIES_H)
+#define KRATOS_FLUID_ADJOINT_UTILITIES_H
+
+// System includes
+#include <array>
+
+// External includes
+
+// Project includes
+#include "geometries/geometry.h"
+#include "includes/node.h"
+#include "includes/ublas_interface.h"
+#include "utilities/coordinate_transformation_utilities.h"
+
+// Application includes
+
+namespace Kratos
+{
+///@addtogroup FluidDynamicsApplication
+///@{
+
+///@name Kratos classes
+///@{
+
+template<unsigned int TDim>
+class KRATOS_API(FLUID_DYNAMICS_APPLICATION) FluidAdjointUtilities
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    using NodeType = Node<3>;
+
+    using IndexType = std::size_t;
+
+    using GeometryType = Geometry<NodeType>;
+
+    using CoordinateTransformationUtilities = CoordinateTransformationUtils<Matrix, Vector, double>;
+
+    ///@}
+    ///@name Static Operations
+    ///@{
+
+    template <class TDataType>
+    static const Variable<double>& GetRelevantVariable(
+        const IndexType DirectionIndex,
+        const Variable<TDataType>& rVariable,
+        const std::array<const Variable<double>*, 3>& rAllVariableComponents);
+
+    template <class TDataType, std::size_t TGradientVariableTotalDimensionality>
+    static std::array<const Variable<double>*, TDim> GetRelevantGradientVariableComponentList(
+        const IndexType DirectionIndex,
+        const Variable<TDataType>& rVariable,
+        const std::array<const Variable<double>*, TGradientVariableTotalDimensionality>& rAllGradientVariableComponents);
+
+    ///@}
+    ///@name Classes
+    ///@{
+
+    template <unsigned int TBlockSize>
+    class KRATOS_API(FLUID_DYNAMICS_APPLICATION) SlipUtilities
+    {
+    public:
```

Comment:
I think you created this extra layer to on one hand avoid instantiating the rotation tool each time it is to be used and to encapsulate all the slip stuff into a single place. I think that the introduced complexity isn't worth it in this case.

Reply:
And as well as to remove code duplication as well.

Reply:
Otherwise, either I have to implement them twice or, to use multiple inheritance :/

Reply:
Why code duplication? Can't you simply include these utilities wherever they are required?

Reply:
done as discussed :)